src/cmclient/cmclient.cpp

```diff
@@ -48,13 +64,193 @@ aos::Error CMClient::InstancesUpdateStatus(const aos::Array<aos::InstanceStatus>
+
+        LOG_DBG() << "READ form channel: " << read;
+
+        if (read < static_cast<int>(sizeof(VchanMessageHeader))) {
+            LOG_ERR() << "Read less than header size: " << read;
+            continue;
+        }
+
+        read = vch_read(&mSMvchanHandler, mReceiveBuffer.Get(), header.dataSize);
+        if (read < static_cast<int>(header.dataSize)) {
+            LOG_ERR() << "Read less than expected message: " << read;
+            continue;
+        }
+
+        SHA256Digest calculatedDigest;
+        auto         err = CalculateSha256(mReceiveBuffer, header.dataSize, calculatedDigest);
+        if (!err.IsNone()) {
+            LOG_ERR() << "Can't calculate SHA256: " << err;
+            continue;
+        }
+
+        if (0 != memcmp(calculatedDigest, header.sha256, sizeof(SHA256Digest))) {
+            LOG_ERR() << "SHA256 mismatch";
+            continue;
+        }
+
+        mIncomingMessage = servicemanager_v3_SMIncomingMessages_init_zero;
+
+        pb_istream_t stream = pb_istream_from_buffer((const uint8_t*)mReceiveBuffer.Get(), header.dataSize);
+        auto         status = pb_decode(&stream, servicemanager_v3_SMIncomingMessages_fields, &mIncomingMessage);
+        if (!status) {
+            LOG_ERR() << "Can't decode incoming PB message: " << PB_GET_ERROR(&stream);
+            continue;
+        }
+
+        switch (mIncomingMessage.which_SMIncomingMessage) {
+        case servicemanager_v3_SMIncomingMessages_get_unit_config_status_tag:
+            break;
+
+        case servicemanager_v3_SMIncomingMessages_check_unit_config_tag:
+            break;
+
+        case servicemanager_v3_SMIncomingMessages_set_unit_config_tag:
+            break;
+
+        case servicemanager_v3_SMIncomingMessages_run_instances_tag:
+            break;
+
+        case servicemanager_v3_SMIncomingMessages_system_log_request_tag:
+            break;
+
+        case servicemanager_v3_SMIncomingMessages_instance_log_request_tag:
+            break;
+
+        case servicemanager_v3_SMIncomingMessages_instance_crash_log_request_tag:
+            break;
+
+        case servicemanager_v3_SMIncomingMessages_image_content_info_tag:
+            break;
+
+        case servicemanager_v3_SMIncomingMessages_image_content_tag:
+            break;
+
+        default:
+            break;
+        }
+    }
+}
+
+void CMClient::SendNodeConfiguration()
+{
+    aos::LockGuard lock(mSendMutex);
 
-    while (true) {
-        LOG_DBG() << "Process message: " << i++;
+    mOutgoingMessage.which_SMOutgoingMessage = servicemanager_v3_SMOutgoingMessages_node_configuration_tag;
+    mOutgoingMessage.SMOutgoingMessage.node_configuration
+        = servicemanager_v3_NodeConfiguration servicemanager_v3_NodeConfiguration_init_zero;
+    auto&                                     config = mOutgoingMessage.SMOutgoingMessage.node_configuration;
 
-        sleep(2);
+    strcpy(config.node_id, CONFIG_NODE_ID);
```

Comment:
Do not use define directly: first it is not recommended by autosar, second not proper c++ way :).
Create `static constexpr auto cNodeType = CONFIG_NODE_TYPE' as CMClient member and use it.

Reply:
done